### PR TITLE
Reland "Speed up first asset load by using the binary-formatted asset manifest for image resolution"

### PR DESCRIPTION
**Pull request by andrewkolos**
Relands https://github.com/flutter/flutter/pull/121322, which causing image loading issues in Wondrous, see https://github.com/flutter/flutter/issues/122447.

It appears the issue was a misplaced cache write in `CachingAssetBundle.loadStrucutredBinaryData` ([code](https://github.com/flutter/flutter/pull/122505/files#diff-1b6c749abe1192b78f5ddab5a43a95e382b6a7d9dc4fd675559d7e4b8800c8da)).

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/services/asset_bundle.dart [hunk 1]
```diff
@@ -266,6 +266,7 @@ abstract class CachingAssetBundle extends AssetBundle {
       .then<T>(parser)
       .then<void>((T value) {
         result = SynchronousFuture<T>(value);
+        _structuredBinaryDataCache[key] = result!;
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/services/asset_bundle.dart:269]: How about a unit test that directly invokes loadStructuredBinaryData and asserts that the cache is populated sychonously?

File: packages/flutter/test/services/asset_bundle_test.dart [hunk 2]
```diff
@@ -178,4 +188,11 @@ void main() {
     expect(assetManifest.keys.toList(), equals(<String>['one']));
     expect(assetManifest['one'], <Object>[]);
   });
+
+  test('loadStructuredBinaryData cache is populated synchronously if load is synchronous', () {
+    final SynchronousTestAssetBundle bundle = SynchronousTestAssetBundle();
+    bundle.loadStructuredBinaryData('one', (ByteData data) => 1);
+    final FutureOr<int> data = bundle.loadStructuredBinaryData('one', (ByteData data) => 0);
+    expect(data, isA<SynchronousFuture<int>>());
+  });
```
Comment (jonahwilliams) [inline packages/flutter/test/services/asset_bundle_test.dart]: How does this fail if your change isn't present?

Comment (andrewkolos) [reply packages/flutter/test/services/asset_bundle_test.dart]: It totally didn't. My bad. Should be fixed now.

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Pull request merged**